### PR TITLE
modularize C source code

### LINKED ISSUE #4339: modularize c source
netdata daemon is modular. However the structure of the source code does not promote this modularity.

To modularize the source code, we need to make significant changes to `automake` and probably move code around. So, I suggest to make these changes once we move to `cmake`.

The C source currently has these plugins:

module|O/S|runtime|type|description
:----:|:---:|:---:|:---:|:---
apps.plugin|linux freebsd|process|data collection|collects metrics from the process tree. It is an independent process because it needs to run with specific capabilities (or as `root`).
freeipmi.plugin|linux|process|data collection|collects enterprise hardware sensors. It is an independent process because it needs to run as `root`.
proc.plugin|linux|thread|data collection|collects system metrics on linux (`/proc`, `/sys` , like `ipc`). This is also modular by itself.
diskspace.plugin|linux|thread|data collection|collects disk space metrics on linux (it is a separate plugin because it requires to be run on its own thread due to the latency of collecting disk space utilization on linux)
freebsd.plugin|freebsd|thread|data collection|collects system metrics on freebsd. This is also modular by itself.
macos.plugin|macos|thread|data collection|collects system metrics on macos. This is also modular by itself.
cgroups.plugin|linux|thread|data collection|collects metrics about containers and VMs. It uses 2 helpers: `cgroup-network` and `cgroup-name`.
nfacct.plugin|linux|thread|data collection|collects netfilter metrics (accounting and connection tracker) using `ibmnl`. Requires `root` access, so it should be converted to a process.
pluginsd.plugin|all|a thread per source|data collection|collects metrics from external plugins and receiving streamed metrics. It spawns a thread for each external plugin found and each remote netdata streaming metrics.
statsd.plugin|all|2 threads|data collection|collects and processes statsd metrics. One thread is the data collector listening on the statsd ports (multiple tcp and udp ports concurrently). Another thread pushes the metrics to netdata (implementing also statsd synthetic charts).
checks.plugin|all|thread|data collection|evaluation of alternative techniques in data collection
idlejitter.plugin|all|thread|data collection|collects metrics related to the CPU latency.
tc.plugin|linux|thread with an external process|data collection|collects traffic QoS metrics on linux, using `tc-qos-helper`. The external `tc-qos-helper` process runs `tc` to export the current statistics of QoS on all interfaces and the internal thread parses the output of `tc` and pushes the metrics to netdata.
backends.plugin|all|thread|backends|The plugin for archiving metrics to time series dbs (`graphite`., `opentsdb`, `json` in 2 forms each). This is modular by itself (so each backend is one of its modules).
allmetrics.plugin|all|inline|api|This plugin exports all metrics in various formats. It has 3 modules: `shell`, `prometheus` and `prometheus_all_hosts`.
registry.plugin|all|inline|api|Implements the netdata registry.
rrdpush.plugin|all|thread|streaming|Streams all metrics to another netdata (they are received by pluginsd.plugin)
health.plugin|all|thread|alarms|Evaluates alarms and sends notifications with `alarm-notify`.

### PR DESCRIPTION
**Pull request by ktsaou**
fixes #4339 

Targets:

1. netdata should still compile and install with `autoconf` and `automake`
2. netdata should compile (no install yet) with `cmake`, using a supplied `config.h`

Data collection plugins:

- [x] `cgroups.plugin` (containers, VMs and `systemd` monitoring on Linux)
- [x] `freeipmi.plugin` (hardware sensors monitoring on Linux)
- [x] `apps.plugin` (monitor the process tree on Linux and FreeBSD)
- [x] `proc.plugin` (monitor linux system metrics)
- [x] `diskspace.plugin` (monitor linux diskspace)
- [x] `checks.plugin` (debugging plugin)
- [x] `freebsd.plugin` (monitor freebsd system metrics)
- [x] `idlejitter.plugin` (monitor cpu latency)
- [x] `nfacct.plugin` (monitor netfilter connection tracker and accounting)
- [x] `tc.plugin` (monitor linux traffic QoS)
- [x] `macos.plugin` (monitor macos system metrics)
- [x] `plugins.d.plugin` (collect metrics from external plugins)
- [x] `statsd.plugin` (collect APM metrics)

netdata internal modules
- [x] `api` (the API of netdata)
- [x] `backends` (metrics archiving to time-series databases)
- [x] `database` (the netdata database)
- [x] `health` (execute health checks on the collected metrics)
- [x] `libnetdata` (a library of code)
- [x] `registry` (the netdata registry)
- [x] `streaming` (metrics streaming and replication)
- [x] `webserver` (metrics streaming and replication)

Problems (all solved):

- [x] ~~For some strange reason, every second compilation, `config.h` get corrupted and compilation fails. This is triggered by the cgroups plugins, because I enabled `.in` processing, since I moved its helpers to the same directory.~~ **FIXED** with workaround.
- [x] ~~on alpine linux `-flto` does not work any more - it seems to be an `ar` plugin problem at the alpine toolchain~~ **FIXED** alpine `ar` and `gcc-ar` are missing the `lto` plugin, so there is no way to fix it with static libraries - so now we use `automake` `subdir-objects` to solve the issue.
- [x] ~~`autoconf` complains that the binaries (apart netdata) reference files in multiple directories. The solution is probably to build `libnetdata.a` with the common files and link all binaries with it.~~ **FIXED**

Problems that will not be solved with this PR:

- modularization means there should be a well defined API for writing plugins. This is true for data collection plugins but it is not for anything else. So, although the plugins are pretty separated by the core, there are a few stupid cases, like `health.plugin`. with circular references (i.e. the core of netdata requires a data structure from the plugin). These issues need some more work to get fixed.

Other notes:

- This PR includes a complete re-work on the `CMakelists.txt`. Now all the compile time features of automake are supported: system libraries are detected, plugins are enabled and disabled based on them, compiler features are detected too (like `-flto`), etc. Though 3 features are still missing:

    - dynamically generate `config.h` (i.e. replace `autoconf`)
    - install netdata (i.e. `make install`)
    - relocating netdata (i.e. compile it and install on custom paths)

### CODE REVIEW

**Comment by ktsaou** [comment]
btw, you can browse this directory structure here: https://github.com/ktsaou/netdata/tree/plugins2/src

**Comment by ktsaou** [comment]
So, in this PR, apart from moving around the files, changing their include files, adding Makefiles to directories and relocating some code to properly compile under this structure... nothing significant has changed.

For `automake` this PR relies on `subdir-objects`, a feature that allows it to use code in multiple directories (btw, `automake` is more efficient than `cmake` in this - code used in multiple binaries, is compiled only once, while `cmake` compiles it for every binary it is used).

No feature changes introduced or changed. So, despite the large number of files changed, this PR should be quite safe to merge.

@vlvkobal @paulfantom @Ferroin please check the new directory structure here: https://github.com/ktsaou/netdata/tree/plugins2/src

I think I should also move `alarm-notify.sh` to `health` and `tc-qos-helper.sh` to `plugins/linux-tc.plugin`.

**Comment by vlvkobal** [comment]
> However, enabling link time optimization on the produced binaries, required the ar command to have the lto plugin.

Is plugin support for `ar` disabled or there is no plugin for lto objects in Alpine Linux?

**Comment by Ferroin** [comment]
Regarding LTO, do we actually gain any tangible benefit from using it?  Put simply, if it's not giving us any visible improvement in performance, and only saves us a few kB of space in the compiled binaries, then I'd argue it's not worth worrying about.

Put differently, seeing some info about build times, final build size, and overall performance both with and without LTO enabled would be kind of nice right now, because if we're not really getting any benefit, there's no point in hobbling the build system just to support it on a small subset of our target platforms.

Beyond that, LGTM.

**Comment by ktsaou** [comment]
LTO is quite important. Netdata has a lot of manual function inlining so generally the benefits of LTO should not be easily visible today. But compile time LTO means we can worry less about it in the future and avoid all this manual work.

To experience the benefits of LTO, you should stress netdata. For example, hit it with thousands of API requests per second so that it saturates the CPU resources. Or collect 1M metrics per second. Or send to it tens of thousands of registry requests per second, Or stream thousands of slaves to it.

On an idle system, with netdata using just 1-2% of a single core, there are a lot of other factors that will influence its performance, so  LTO benefits will not be visible.

**Comment by ktsaou** [comment]
> https://stackoverflow.com/questions/5091583/cmake-reuse-object-files-built-for-a-lib-into-another-lib-target

nice finding!
Used that feature now, so `cmake` is as efficient as `automake`.

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,269 +3,514 @@
...
+set(NETDATA_COMMON_CFLAGS ${NETDATA_COMMON_CFLAGS} ${ZLIB_CFLAGS_OTHER})
+set(NETDATA_COMMON_LIBRARIES ${NETDATA_COMMON_LIBRARIES} ${ZLIB_LIBRARIES})
+set(NETDATA_COMMON_INCLUDE_DIRS ${NETDATA_COMMON_INCLUDE_DIRS} ${ZLIB_INCLUDE_DIRS})
+
+
+# -----------------------------------------------------------------------------
+# Detect libcap
+
+IF(LINUX)
+    pkg_check_modules(CAP QUIET libcap)
```
Comment (vlvkobal) [inline CMakeLists.txt]: FreeBSD has built-in support for [Capabilities](https://www.freebsd.org/cgi/man.cgi?query=capsicum&sektion=4&apropos=0&manpath=FreeBSD+11.2-RELEASE+and+Ports) and [POSIX.1e](https://www.freebsd.org/cgi/man.cgi?query=posix1e&apropos=0&sektion=0&manpath=FreeBSD+11.2-RELEASE+and+Ports&arch=default&format=html)

Comment (ktsaou) [reply CMakeLists.txt]: Is that compatible with `libcap` under linux? `apps.plugin` uses Linux capabilities instead of `setuid` to root.

Comment (vlvkobal) [reply CMakeLists.txt]: No. The API is different.

Comment (ktsaou) [reply CMakeLists.txt]: So, we don't support it yet.

**Comment by ktsaou** [comment]
I also tested on an ancient ubuntu 8.04 (it was quite a trick to install this version in a VM).
This PR compiles and runs fine (though `pkg-config` is broken on these systems and requires running the installer with `--libs-are-really-here`).

I also tested `arch`, `gentoo`, `ubuntu-18.04`, `ubuntu-12.04`, `rhel7`, `rhel6`.

**Comment by paulfantom** [comment]
> I also tested on an ancient ubuntu 8.04 

WHY? This system is over 10yo, no one gives more that 10yo of support, if someone still uses such ancient and unsupported system then it is his problem not ours.

**Comment by ktsaou** [comment]
> WHY? This system is over 10yo, no one gives more than 10yo of support, if someone still uses such ancient and unsupported system then it is his problem not ours.

I strongly believe time will go back one day... :)
Just kidding...

If it works on such an ancient system and at the same time works on a fully updated ArchLinux, then it most probably works everywhere...

Actually, I wanted to test a system from 2003-2004. But I don't have one handy...

**Comment by paulfantom** [comment]
RHEL 2.1 (2002): http://legacy.redhat.com/pub/redhat/rhdb/2.1/en/iso/i386/ Good Luck.

**Pull request merged**